src/jit/lower.cpp

```diff
@@ -571,6 +579,240 @@ void Lowering::DecomposeNode(GenTreePtr* pTree, Compiler::fgWalkData* data)
+
+    GenTree* addrBase = tree->gtOp.gtOp1;
+    GenTree* dataHigh = gtLong->gtOp.gtOp2;
+    GenTree* dataLow = gtLong->gtOp.gtOp1;
+    GenTree* storeIndLow = tree;
+
+    // Rewrite storeIndLow tree to save only lower 32-bit data.
+    // 
+    //  |  |  {  |  /--*  lclVar    byref  V24 rat2   (address)
+    //  ...
+    //  |  |  {  |  +--*  lclVar    int    V25 rat3   (lower 32-bit data)
+    //  |  |  {  |  {  *  stmtExpr  void  (embedded) (IL   ???...  ???)
+    //  |  |  {  |  {  |  /--*  lclVar    int    V23 rat1
+    //  |  |  {  |  {  \--*  st.lclVar int    V26 rat4
+    //  |  |  {  \--*  storeIndir int
+    comp->fgSnipNode(currStmt, gtLong);
+    comp->fgSnipNode(currStmt, dataHigh);
+    storeIndLow->gtOp.gtOp2 = dataLow;
+    storeIndLow->gtType = TYP_INT;
+
+    // Construct storeIndHigh tree
+    //
+    // | | {  *stmtExpr  void  (embedded)(IL ? ? ? ... ? ? ? )
+    // | | { | / --*  lclVar    int    V26 rat4
+    // | | { | | / --*  lclVar    byref  V24 rat2
+    // | | { | +--*  lea(b + 4)  ref
+    // | | {  \--*  storeIndir int
+    GenTree* addrBaseHigh = new(comp, GT_LCL_VAR) GenTreeLclVar(GT_LCL_VAR, 
+        addrBase->TypeGet(), addrBase->AsLclVarCommon()->GetLclNum(), BAD_IL_OFFSET);
+    GenTree* addrHigh = new(comp, GT_LEA) GenTreeAddrMode(TYP_REF, addrBaseHigh, nullptr, 0, genTypeSize(TYP_INT));
+    GenTree* storeIndHigh = new(comp, GT_STOREIND) GenTreeStoreInd(TYP_INT, addrHigh, dataHigh);
+    storeIndHigh->gtFlags = storeIndLow->gtFlags;
+    storeIndHigh->gtFlags |= GTF_REVERSE_OPS;
+    storeIndHigh->CopyCosts(storeIndLow);
+
+    // External links of storeIndHigh tree
+    dataHigh->gtPrev = nullptr;
+    if (isEmbeddedStmt)
+    {
+        // If storeIndTree is an embedded statement, connect storeIndLow
+        // and dataHigh
+        storeIndLow->gtNext = dataHigh;
+        dataHigh->gtPrev = storeIndLow;
+    }
+    storeIndHigh->gtNext = linkEnd;
+    if (linkEnd)
+    {
+        linkEnd->gtPrev = storeIndHigh;
+    }
+
+    // Internal links of storeIndHigh tree
+    dataHigh->gtNext = addrBaseHigh;
+    addrBaseHigh->gtPrev = dataHigh;
+    addrBaseHigh->gtNext = addrHigh;
+    addrHigh->gtPrev = addrBaseHigh;
+    addrHigh->gtNext = storeIndHigh;
+    storeIndHigh->gtPrev = addrHigh;
+    
+
+    if (isEmbeddedStmt)
+    {
+        // Find a parent statment containing storeIndHigh. 
+        GenTree* parentStmt = currStmt;
+        while (parentStmt && !parentStmt->AsStmt()->gtStmtIsTopLevel())
+        {
+            parentStmt = parentStmt->gtPrev;
+        }
+        assert(parentStmt);
+
+        GenTreeStmt* stmt = comp->fgMakeEmbeddedStmt(comp->compCurBB, storeIndHigh, parentStmt);
+        stmt->gtStmtILoffsx = comp->compCurStmt->gtStmt.gtStmtILoffsx;
+    }
+    else
+    {
+        GenTreeStmt* stmt = comp->fgNewStmtFromTree(storeIndHigh);
+        stmt->gtStmtILoffsx = comp->compCurStmt->gtStmt.gtStmtILoffsx;
+
+        // Find an insert point. Skip all embedded sattements created above.
```

Comment:
sattements -> statements

Also, embedded statements may already exists so this isn't only about the ones created above.